staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -366,7 +371,26 @@ func ValidateCustomResourceDefinitionSpecUpdate(spec, oldSpec *apiextensions.Cus
 	// find out whether any schema had default before. Then we keep allowing it.
 	allowDefaults := utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceDefaulting) || specHasDefaults(oldSpec)
 
-	allErrs := validateCustomResourceDefinitionSpec(spec, requireRecognizedVersion, allowDefaults, fldPath)
+	mungedCRDSpec := spec.DeepCopy()
+	if apiextensions.HasPerVersionSchema(mungedCRDSpec.Versions) {
+		// drop invalid type for per-version schemas
+		for i, mungedCRDVersion := range mungedCRDSpec.Versions {
+			for j, oldCRDVersion := range oldSpec.Versions {
+				if mungedCRDVersion.Name == oldCRDVersion.Name {
+					if mungedCRDVersion.Schema != nil && oldCRDVersion.Schema != nil {
+						dropInvalidJSONSchemaFields(mungedCRDSpec.Versions[i].Schema.OpenAPIV3Schema, oldSpec.Versions[j].Schema.OpenAPIV3Schema)
```

Comment:
mutating a copy of the spec inside validation is not what I expected. the changes I expected:

1. do not publish openapi for a schema containing bad types
2. do not change validation (yet)
3. in v1 validation only (https://github.com/kubernetes/kubernetes/pull/79604) require valid object types on create, and do not allow changing valid to invalid on update

Reply:
for publishing openapi we could do something like https://github.com/roycaihw/kubernetes/commit/2a75b8e8fabc8eeaf72b6808714fb536e81a3269. It's probably better to do the validation on JSONSchemaProps (like what you have here) instead of structural schema so we can save some code between schema validation and publishing

Reply:
>  It's probably better to do the validation on JSONSchemaProps (like what you have here) instead of structural schema

validating structural really saves code comparing to JSONSchemaProps at least at this point, b/c the json model is far more complex and most fields are even unused/unsupported actually. i just tried another approach basing on your commit.

Reply:
I had `old *Structural` in the interface because I had [2 & 3](https://github.com/kubernetes/kubernetes/pull/79587#discussion_r300005679) in my mind. For openapi publishing old is always nil (treated as CREATE)

Reply:
> I had 2 & 3 in my mind. For openapi publishing old is always nil (treated as CREATE)

oh thanks i mis-understood. if we will apply the validation in v1, it should better be validating JSONSchemaProps, isn't it?

Reply:
yes. JSONSchemaProps needs a backwards-compatible validation in v1; in v1beta1, the openapi controller can simply ignore invalid schemas by validating either JSONSchemaProps or Structural type